[clang-format] Upgrade ShortFunctionStyle to a struct

irymarchyk opened the pull request:
The current clang-format configuration option AllowShortFunctionsOnASingleLine uses a single enum (ShortFunctionStyle) to control when short function definitions can be merged onto a single line. This enum provides predefined combinations of conditions (e.g., None, Empty only, Inline only, Inline including Empty, All).

This approach has limitations:

1. **Lack of Granularity:** Users cannot specify arbitrary combinations of conditions. For example, a user might want to allow merging for both empty functions and short top-level functions, but not for short functions defined within classes. This is not possible with the current enum options except by choosing All, which might merge more than desired.

2. **Inflexibility:** Adding new conditions for merging (e.g., distinguishing between member functions and constructors, handling lambdas specifically) would require adding many new combined enum values, leading to a combinatorial explosion and making the configuration complex.

3. **Implicit Behavior:** Some options imply others (e.g., Inline implies Empty), which might not always be intuitive or desired.

The goal is to replace this single-choice enum with a more flexible mechanism allowing users to specify a set of conditions that must be met for a short function to be merged onto a single line.

**Proposed Solution**

1. Introduce a new configuration option: AllowShortFunctionsOnSingleLineOptions.

2. This option will accept a list of strings, where each string represents a specific condition allowing merging.

3. **Backward Compatibility:**

    - If AllowShortFunctionsOnSingleLineOptions is present in the configuration, it takes precedence.

    - If AllowShortFunctionsOnSingleLineOptions is not present, but the old AllowShortFunctionsOnASingleLine is present, the old option should be parsed and mapped to the corresponding new semantics for compatibility.

Commit:
[clang-format]: Add `Custom` to `ShortFunctionStyle`; add new AllowShortFunctionsOnASingleLineOptions for granular setup

The current clang-format configuration option AllowShortFunctionsOnASingleLine uses a single enum (ShortFunctionStyle) to control when short function definitions can be merged onto a single line. This enum provides predefined combinations of conditions (e.g., None, Empty only, Inline only, Inline including Empty, All).

This approach has limitations:

1. **Lack of Granularity:** Users cannot specify arbitrary combinations of conditions. For example, a user might want to allow merging for both empty functions and short top-level functions, but not for short functions defined within classes. This is not possible with the current enum options except by choosing All, which might merge more than desired.

2. **Inflexibility:** Adding new conditions for merging (e.g., distinguishing between member functions and constructors, handling lambdas specifically) would require adding many new combined enum values, leading to a combinatorial explosion and making the configuration complex.

3. **Implicit Behavior:** Some options imply others (e.g., Inline implies Empty), which might not always be intuitive or desired.

The goal is to replace this single-choice enum with a more flexible mechanism allowing users to specify a set of conditions that must be met for a short function to be merged onto a single line.

**Proposed Solution**

1. Introduce a new configuration option: AllowShortFunctionsOnSingleLineOptions.

2. This option will accept a list of strings, where each string represents a specific condition allowing merging.

3. **Backward Compatibility:**

    - If AllowShortFunctionsOnSingleLineOptions is present in the configuration, it takes precedence.

    - If AllowShortFunctionsOnSingleLineOptions is not present, but the old AllowShortFunctionsOnASingleLine is present, the old option should be parsed and mapped to the corresponding new semantics for compatibility.

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -871,13 +871,81 @@ struct FormatStyle {
+    ///   };
+    ///   void f() { bar(); }
+    /// \endcode
+    bool All;
```
```suggestion
    bool Other;
```
Below is the equivalence table based on my understanding:
| `enum` | `Custom` |
| --- | --- |
| `SFS_None` | everything set to `false` |
| `SFS_InlineOnly` | only `Inline` set to `true` |
| `SFS_Empty` | only `Empty` set to `true` |
| `SFS_Inline` | both `Inline` and `Empty` set to `true` |
| `SFS_All` | everything set to `true` |

irymarchyk replied:
That was my understanding too, but I have to set only 'All/Other' to true when SFS_All was specified to pass unit tests. 
There seems to be JS dependency on SFS_Inline/SFS_InlineOnly:
https://github.com/llvm/llvm-project/blob/78905ce6cbd3fa8f8b467e7cad0e9a093c1b1c44/clang/lib/Format/TokenAnnotator.cpp#L5697
(in my PR it was changed a little bit)

And this is unit test which fails if I set every boolean to true when SFS_All is set:
https://github.com/llvm/llvm-project/blob/78905ce6cbd3fa8f8b467e7cad0e9a093c1b1c44/clang/unittests/Format/FormatTestJS.cpp#L1141

Seems like the code only works for Inline cases, but not for SFS_All case... So I have to set Inline to false when SFS_All (and Empty too to be consistent).
```
/Users/irymarchyk/dev/llvm-project/clang/unittests/Format/FormatTestJS.cpp:51: Failure
Expected equality of these values:
  Code.str()
    Which is: "var func = function() { return 1; };"
  format(Code, Style)
    Which is: "var func = function() {\n  return 1;\n};"
With diff:
@@ -1,1 +1,3 @@
-var func = function() { return 1; };
+var func = function() {
+  return 1;
+};

Expected code is not stable

/Users/irymarchyk/dev/llvm-project/clang/unittests/Format/FormatTestJS.cpp:53: Failure
Expected equality of these values:
  Code.str()
    Which is: "var func = function() { return 1; };"
  Result
    Which is: "var func = function() {\n  return 1;\n};"
With diff:
@@ -1,1 +1,3 @@
-var func = function() { return 1; };
+var func = function() {
+  return 1;
+};

Formatted:
var func = function() {
  return 1;
};

/Users/irymarchyk/dev/llvm-project/clang/unittests/Format/FormatTestJS.cpp:51: Failure
Expected equality of these values:
  Code.str()
    Which is: "var outer = function() { var inner = function() { return 1; } };"
  format(Code, Style)
    Which is: "var outer = function() {\n  var inner = function() { return 1; }\n};"
With diff:
@@ -1,1 +1,3 @@
-var outer = function() { var inner = function() { return 1; } };
+var outer = function() {
+  var inner = function() { return 1; }
+};

Expected code is not stable

/Users/irymarchyk/dev/llvm-project/clang/unittests/Format/FormatTestJS.cpp:53: Failure
Expected equality of these values:
  Code.str()
    Which is: "var outer = function() { var inner = function() { return 1; } };"
  Result
    Which is: "var outer = function() {\n  var inner = function() { return 1; }\n};"
With diff:
@@ -1,1 +1,3 @@
-var outer = function() { var inner = function() { return 1; } };
+var outer = function() {
+  var inner = function() { return 1; }
+};

Formatted:
var outer = function() {
  var inner = function() { return 1; }
};

[  FAILED  ] FormatTestJS.InliningFunctionLiterals (24 ms)

```

owenca replied:
See https://github.com/llvm/llvm-project/pull/134337#discussion_r2029943820.

irymarchyk replied:
Thanks, now the issue is resolved.

Commit:
PR-134337: fixed PR comments

irymarchyk commented on clang/include/clang/Format/Format.h:
```diff
@@ -871,13 +871,81 @@ struct FormatStyle {
+  ///     All: false
+  /// \endcode
+  struct ShortFunctionMergeFlags {
+    /// Only merge empty functions.
```
Yes, it is covering these. I've added test case for this:
```
  // test with comment
  verifyFormat("void f3() { /* comment */ }", CustomEmpty);
```
I fixed comment

irymarchyk commented on clang/include/clang/Format/Format.h:
```diff
@@ -871,13 +871,81 @@ struct FormatStyle {
+    /// Only merge functions defined inside a class.
+    /// \code
+    ///   class Foo {
+    ///     void f() { foo(); }
```
Thanks, fixed

irymarchyk commented on clang/include/clang/Format/Format.h:
```diff
@@ -871,13 +871,81 @@ struct FormatStyle {
+    ///   }
+    /// \endcode
+    bool Empty;
+    /// Only merge functions defined inside a class.
```
Thanks, fixed

irymarchyk commented on clang/include/clang/Format/Format.h:
```diff
@@ -871,13 +871,81 @@ struct FormatStyle {
+    ///   void f() {
+    ///     foo();
+    ///   }
+    ///   void f() {}
```
Thanks, I've copy/pasted and somehow missed this. Unit-test has this case and it works as expected (multiline)

owenca commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -5687,11 +5687,10 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
              (Left.NestingLevel == 0 && Line.Level == 0 &&
-              Style.AllowShortFunctionsOnASingleLine &
-                  FormatStyle::SFS_InlineOnly);
+              Style.AllowShortFunctionsOnASingleLineOptions.Inline);
```
```suggestion
              Style.AllowShortFunctionsOnASingleLineOptions.Inline &&
              !Style.AllowShortFunctionsOnASingleLineOptions.Other);
```
`& SFS_InlineOnly` was a shorthand for `SFS_Inline || SFS_InlineOnly`, so it seems that you missed `Other`.

owenca commented on clang/lib/Format/Format.cpp:
```diff
@@ -1472,6 +1483,30 @@ static void expandPresetsSpacesInParens(FormatStyle &Expanded) {
+  // Reset all flags
+  Expanded.AllowShortFunctionsOnASingleLineOptions = {};
+
+  if (Expanded.AllowShortFunctionsOnASingleLine == FormatStyle::SFS_None)
+    return;
+
+  if (Expanded.AllowShortFunctionsOnASingleLine == FormatStyle::SFS_Empty ||
+      Expanded.AllowShortFunctionsOnASingleLine == FormatStyle::SFS_Inline) {
+    Expanded.AllowShortFunctionsOnASingleLineOptions.Empty = true;
+  }
+
+  if (Expanded.AllowShortFunctionsOnASingleLine == FormatStyle::SFS_Inline ||
+      Expanded.AllowShortFunctionsOnASingleLine ==
+          FormatStyle::SFS_InlineOnly) {
+    Expanded.AllowShortFunctionsOnASingleLineOptions.Inline = true;
+  }
+
+  if (Expanded.AllowShortFunctionsOnASingleLine == FormatStyle::SFS_All)
+    Expanded.AllowShortFunctionsOnASingleLineOptions.Other = true;
```
```suggestion
  switch (Expanded.AllowShortFunctionsOnASingleLine) {
  case FormatStyle::SFS_All:
    Expanded.AllowShortFunctionsOnASingleLineOptions.Inline = true;
    Expanded.AllowShortFunctionsOnASingleLineOptions.Other = true;
    [[fallthrough]];
  case FormatStyle::SFS_Empty:
    Expanded.AllowShortFunctionsOnASingleLineOptions.Empty = true;
    break;
  case FormatStyle::SFS_Inline:
    Expanded.AllowShortFunctionsOnASingleLineOptions.Empty = true;
    [[fallthrough]];
  case FormatStyle::SFS_InlineOnly:
    Expanded.AllowShortFunctionsOnASingleLineOptions.Inline = true;
  }
    break;
  default:
    break;
```
Use `switch` like other expansions.

irymarchyk replied:
This is not relevant anymore.

owenca commented on clang/lib/Format/Format.cpp:
```diff
@@ -1501,6 +1536,8 @@ FormatStyle getLLVMStyle(FormatStyle::LanguageKind Language) {
   LLVMStyle.AllowShortCompoundRequirementOnASingleLine = true;
   LLVMStyle.AllowShortEnumsOnASingleLine = true;
   LLVMStyle.AllowShortFunctionsOnASingleLine = FormatStyle::SFS_All;
+  LLVMStyle.AllowShortFunctionsOnASingleLineOptions = {};
```
```suggestion
  LLVMStyle.AllowShortFunctionsOnASingleLineOptions.Empty = true;
  LLVMStyle.AllowShortFunctionsOnASingleLineOptions.Inline = true;
```
`SFS_All` requires setting everything.

irymarchyk replied:
Thanks, fixed!

owenca commented on clang/lib/Format/Format.cpp:
```diff
@@ -1893,6 +1938,8 @@ FormatStyle getChromiumStyle(FormatStyle::LanguageKind Language) {
   } else {
     ChromiumStyle.AllowAllParametersOfDeclarationOnNextLine = false;
     ChromiumStyle.AllowShortFunctionsOnASingleLine = FormatStyle::SFS_Inline;
+    ChromiumStyle.AllowShortFunctionsOnASingleLineOptions = {};
+    ChromiumStyle.AllowShortFunctionsOnASingleLineOptions.Inline = true;
```
```suggestion
    ChromiumStyle.AllowShortFunctionsOnASingleLineOptions.Empty = true;
     ChromiumStyle.AllowShortFunctionsOnASingleLineOptions.Inline = true;
    ChromiumStyle.AllowShortFunctionsOnASingleLineOptions.Other = false;
```
`SFS_Inline` requires setting both `Empty` and `Inline`.

irymarchyk replied:
Thanks, fixed!

owenca commented on clang/lib/Format/Format.cpp:
```diff
@@ -1907,6 +1954,8 @@ FormatStyle getMozillaStyle() {
   FormatStyle MozillaStyle = getLLVMStyle();
   MozillaStyle.AllowAllParametersOfDeclarationOnNextLine = false;
   MozillaStyle.AllowShortFunctionsOnASingleLine = FormatStyle::SFS_Inline;
+  MozillaStyle.AllowShortFunctionsOnASingleLineOptions = {};
+  MozillaStyle.AllowShortFunctionsOnASingleLineOptions.Inline = true;
```
Ditto.

irymarchyk replied:
Thanks, fixed!

sstwcw commented:
When I changed the `AlignConsecutiveAssignments` option, I was told to reuse the same option name.  Now the program allows the user to specify the option as either a string or an associative array.

owenca commented:
> When I changed the `AlignConsecutiveAssignments` option, I was told to reuse the same option name. Now the program allows the user to specify the option as either a string or an associative array.

Reusing the option name is IMO better than adding `SFS_Custom`. @irymarchyk is it doable?

irymarchyk commented:
If this was done previously I think this is doable. I will take a look and make appropriate changes. Thanks.

Commit:
Updated configuration with new suggestion - reuse existing option

irymarchyk commented on clang/lib/Format/Format.cpp:
```diff
@@ -1500,7 +1523,10 @@ FormatStyle getLLVMStyle(FormatStyle::LanguageKind Language) {
   LLVMStyle.AllowShortEnumsOnASingleLine = true;
-  LLVMStyle.AllowShortFunctionsOnASingleLine = FormatStyle::SFS_All;
+  LLVMStyle.AllowShortFunctionsOnASingleLine =
+      FormatStyle::ShortFunctionStyle({/*Empty=*/true,
```
Please let me know if it is acceptable to make static functions like `FormatStyle::ShortFunctionStyle::SFS_Empty()` (or similar) so we don't have specify each member in constructor every time. This might be useful in the future - I want to add few options to ShortFunctionStyle, and I have to increase number of parameters in constructor. This will require changes in every place where it was used.

irymarchyk replied:
I've made changes; please take a look.

Commit:
fixed documentation error

sstwcw commented on clang/docs/ClangFormatStyleOptions.rst:
```diff
@@ -1959,6 +1964,52 @@ the configuration (without a prefix: ``Auto``).
+    AllowShortFunctionsOnASingleLine:
+      Empty: false
+      Inline: true
+      All: false
```
All should probably be other.

irymarchyk replied:
Thanks, fixed

sstwcw commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -5687,11 +5687,11 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
-              Style.AllowShortFunctionsOnASingleLine &
-                  FormatStyle::SFS_InlineOnly);
+              Style.AllowShortFunctionsOnASingleLine.Inline &&
+              !Style.AllowShortFunctionsOnASingleLine.Other);
```
It looks like you should not add inline here.  The old check uses `&`.  That is same as comparing against `SFS_All`.  The idea is that the brace should be on a new line if either the `AllosShortFunctionsOnASingleLine` is completely disabled, or the function is at the top level and the other option is disabled.

sstwcw replied:
Since the options are now independent, this part should probably become something like this.

```C++
!(Left.NestingLevel == 0 && Line.Level == 0 ? Style.AllowShortFunctionsOnASingleLine.Other : Style.AllowShortFunctionsOnASingleLine.Inline)
```

irymarchyk replied:
Thanks, fixed.

owenca replied:
> It looks like you should not add inline here. The old check uses `&`. That is same as comparing against `SFS_All`.

See https://github.com/llvm/llvm-project/pull/134337#discussion_r2029943820.

owenca replied:
> Since the options are now independent, this part should probably become something like this.
> 
> ```c++
> !(Left.NestingLevel == 0 && Line.Level == 0 ? Style.AllowShortFunctionsOnASingleLine.Other : Style.AllowShortFunctionsOnASingleLine.Inline)
> ```

That seems wrong. See https://github.com/llvm/llvm-project/pull/134337#discussion_r2043562759.

Commit:
Fixed PR comments

irymarchyk commented on clang/docs/ClangFormatStyleOptions.rst:
```diff
@@ -1959,6 +1964,52 @@ the configuration (without a prefix: ``Auto``).
+      void f() {
+      }
+
+  * ``bool Other`` Merge all functions fitting on a single line. Please note that this
```
Should we expose this option to user? I am not sure what it means if we set only Other, but not others - then C/C++ code will behave like SFS_All. Probably we should leave it to only set it when user sets 'All' in configuration.

irymarchyk commented:
Sorry, I forgot to push latest changes

owenca commented:
We should stick to the table in https://github.com/llvm/llvm-project/pull/134337#discussion_r2028339995 and deprecate `SFS_InlineOnly`, `SFS_Empty`, and `SFS_Inline`.

Commit:
marked options as deprecated

irymarchyk commented:
> We should stick to the table in [#134337 (comment)](https://github.com/llvm/llvm-project/pull/134337#discussion_r2028339995) and deprecate `SFS_InlineOnly`, `SFS_Empty`, and `SFS_Inline`.

Thanks, I marked options as 'deprecated'. Mapping is done in Format.cpp.

owenca commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -293,15 +293,14 @@ class LineJoiner {
 
-      if (Style.AllowShortFunctionsOnASingleLine &
-          FormatStyle::SFS_InlineOnly) {
+      if (Style.AllowShortFunctionsOnASingleLine.Inline) {
```
See https://github.com/llvm/llvm-project/pull/134337#discussion_r2029943820.

irymarchyk replied:
I think because we have check for All in the beginning, we don't need to add another check !All here as it always be true.

irymarchyk replied:
Thanks, fixed. Probably will need to add new unit test to cover this situation in the future.

irymarchyk commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -5687,11 +5687,9 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
-             (Left.NestingLevel == 0 && Line.Level == 0 &&
-              Style.AllowShortFunctionsOnASingleLine &
-                  FormatStyle::SFS_InlineOnly);
+      return !(Left.NestingLevel == 0 && Line.Level == 0
+                   ? Style.AllowShortFunctionsOnASingleLine.Other
+                   : Style.AllowShortFunctionsOnASingleLine.Inline);
```
Sorry, I am confused now with these changes. 
Previously we had setting SFS_All which means 'everything'. Now we introduced option 'Other' and I cannot create explanation for what it does if user sets only it. (Inline is false, Empty is false, Other is true). The only place it is being used is this code (by used I mean has effect on the result). 
I tried to create table to get the difference between old values and new values:
| `enum`           | `Custom`                                | Left.NestingLevel == 0 && Line.Level == 0 | old result | New result |
| ---------------- | --------------------------------------- | ----------------------------------------- | ---------- | ---------- |
| `SFS_None`       | everything set to `false`               | true                                      | true       | true       |
|                  |                                         | false                                     | true       | true       |
| `SFS_InlineOnly` | only `Inline` set to `true`             | true                                      | true       | true       |
|                  |                                         | false                                     | false      | false      |
| `SFS_Empty`      | only `Empty` set to `true`              | true                                      | true       | true       |
|                  |                                         | false                                     | true       | true       |
| `SFS_Inline`     | both `Inline` and `Empty` set to `true` | true                                      | true       | true       |
|                  |                                         | false                                     | false      | false      |
| `SFS_All`        | everything set to `true`                | true                                      | false      | false      |
|                  |                                         | false                                     | false      | false      |
| Other = true     | Other = true                            | true                                      | -          | false      |
|                  |                                         | false                                     | -          | true       |

So new result with only Other==true does not correlate with anything that was before...
Per my understanding, this value (Other) should only be set to true if SFS_All option was specified. In other cases 'Other' does not make sense. I think we should not allow user to set it directly (only through SFS_All). Please let me know what do you think. In this case code can be simplified.

owenca replied:
The new struct is a partition of all cases, with `other` being "everything else", i.e., any cases not covered by `Empty` (or `SFS_Empty` before), `Inline` (or `SFS_InlineOnly` before), and any future struct members. This is similar to `Other` of [`SpacesInParensOptions`](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#spacesinparensoptions).

owenca replied:
If only `Other` is set, we should get the following format:
```
void topLevelEmpty() {
}

void topLevelNonEmpty() { return; }

struct Foo {
  void inlineEmpty() {
  }

  void inlineNonEmpty() {
    return;
  }
};
```

irymarchyk replied:
So is it OK that it won't have any special case (for now) except cases previously covered by SFS_All?
I tried to fix other comments too; please take a look (mostly replaced Other with isAll() call which equals to Empty && Inline && Other)

irymarchyk replied:
Thanks for explanation! I think I finally understood what Other mean. 
I've added your example to unit test.

Commit:
fixed PR comments

Commit:
added usage of 'other'

owenca commented:
@irymarchyk can you resolve the conflicts?

Commit:
Merge remote-tracking branch 'origin/main' into add_custom_AllowShortFunctionsOnASingleLine

irymarchyk commented:
> @irymarchyk can you resolve the conflicts?

Thanks, I hadn't noticed that. Now PR does not have conflicts.

owenca commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -5685,11 +5685,9 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
-              Style.AllowShortFunctionsOnASingleLine &
-                  FormatStyle::SFS_InlineOnly);
+      return !(Left.NestingLevel == 0 && Line.Level == 0
+                   ? Style.AllowShortFunctionsOnASingleLine.isAll()
```
Looks like this is even worse than what @sstwcw suggested because it also seems incorrect if the following is true:
```
Left.NestingLevel == 0 && Line.Level == 0 && Other && !Inline
```

irymarchyk replied:
Thanks, I tried to add code to correctly handle this case. Also I added unit test (per my understanding). Please let me know if you want to add more cases (or if I made some mistake). Thanks.

Commit:
fixed ShortFunctionStyle.Other case for JS

irymarchyk commented:
@owenca can you please review one more time?

owenca commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -291,54 +291,59 @@ class LineJoiner {
+        const AnnotatedLine *Line = nullptr;
+        for (auto J = I - 1; J >= AnnotatedLines.begin(); --J) {
+          assert(*J);
+          if ((*J)->InPPDirective || (*J)->isComment() ||
+              (*J)->Level > TheLine->Level) {
+            continue;
           }
+          if ((*J)->Level < TheLine->Level ||
+              (Style.BreakBeforeBraces == FormatStyle::BS_Whitesmiths &&
+               (*J)->First->is(tok::l_brace))) {
+            Line = *J;
+            break;
+          }
+        }
 
-          if (!Line)
-            return false;
+        if (!Line)
+          return false;
 
-          // Check if the found line starts a record.
-          const auto *LastNonComment = Line->getLastNonComment();
-          // There must be another token (usually `{`), because we chose a
-          // non-PPDirective and non-comment line that has a smaller level.
-          assert(LastNonComment);
-          return isRecordLBrace(*LastNonComment);
-        }
+        // Check if the found line starts a record.
+        const auto *LastNonComment = Line->getLastNonComment();
+        // There must be another token (usually `{`), because we chose a
+        // non-PPDirective and non-comment line that has a smaller level.
+        assert(LastNonComment);
+        return isRecordLBrace(*LastNonComment);
       }
 
-      return false;
+      if (empty_function && !Style.AllowShortFunctionsOnASingleLine.Empty)
+        return false;
+
+      return Style.AllowShortFunctionsOnASingleLine.Other;
```
Is this for the new case where only `Other` is turned on? Can we handle it in a future patch?

irymarchyk replied:
Yes, this is to correctly handle case when 'only Other' is enabled. I've removed this and will add it in subsequent patch. Thanks. (also removed test case related to it)

owenca reviewed (commented):
If there are any new test cases including those for only `Other == true`, can you move them to a future patch?

Commit:
fixed first comment - simplified code

Commit:
removed code related to Only Other option

irymarchyk commented on clang/lib/Format/TokenAnnotator.cpp:
```diff
@@ -5685,11 +5685,27 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
-             (Left.NestingLevel == 0 && Line.Level == 0 &&
-              Style.AllowShortFunctionsOnASingleLine &
-                  FormatStyle::SFS_InlineOnly);
+      const auto &shortFuncConfig = Style.AllowShortFunctionsOnASingleLine;
+
+      // SFS_All
+      if (shortFuncConfig.isAll())
+        return false;
+
+      // SFS_None and SFS_Empty
+      if (shortFuncConfig == FormatStyle::ShortFunctionStyle{})
+        return true;
+
+      // SFS_Empty
+      if (shortFuncConfig == FormatStyle::ShortFunctionStyle{/*Empty=*/true,
+                                                             /*Inline=*/false,
+                                                             /*Other=*/false}) {
+        return true;
+      }
+
+      if (Left.NestingLevel == 0 && Line.Level == 0)
+        return shortFuncConfig.Inline && !shortFuncConfig.Other;
+
+      return shortFuncConfig.Other;
```
Thanks, that's much better.

Commit:
fixed PR comments

Commit:
added one forgotten change

irymarchyk commented on .gitignore:
```diff
@@ -73,3 +73,4 @@ pythonenv*
 /clang/utils/analyzer/projects/*/RefScanBuildResults
 # automodapi puts generated documentation files here.
 /lldb/docs/python_api/
+clang-build/
```
Sorry, pushed by mistake; removed

irymarchyk commented on clang/lib/Format/Format.cpp:
```diff
@@ -622,15 +622,38 @@ template <> struct ScalarEnumerationTraits<FormatStyle::ShortBlockStyle> {
+                FormatStyle::ShortFunctionStyle({/*Empty=*/true,
+                                                 /*Inline=*/true,
+                                                 /*Other=*/true}));
+    IO.enumCase(Value, "false", FormatStyle::ShortFunctionStyle({}));
```
They were mapped before, and there is a unit test which cover these old cases... So I think we still need them.

irymarchyk committed:
Merge branch 'main' into add_custom_AllowShortFunctionsOnASingleLine

owenca committed:
Merge branch 'main' into add_custom_AllowShortFunctionsOnASingleLine

owenca committed:
Refactor ShortFunctionStyle initialization

owenca committed:
Update fmt.Println message from 'Hello' to 'Goodbye'

owenca committed:
Update print statement from 'Hello' to 'Goodbye'

owenca committed:
Refactor AllowShortFunctionsOnASingleLine style usage

owenca committed:
Update print statement from 'Hello' to 'Goodbye'

owenca committed:
Refactor FormatTestJS by cleaning up function styles

Removed redundant initialization of AllowShortFunctionsOnASingleLine and deleted InliningFunctionLiteralsNew test case.

owenca committed:
Update print statement from 'Hello' to 'Goodbye'

owenca committed:
Add tests for AllowShortFunctionsOnASingleLine
